File: whizard/vamp/src/linalg.f90
! linalg.f90 --
! Copyright (C) 1998 by Thorsten Ohl <ohl@hep.tu-darmstadt.de>
! 
! VAMP is free software; you can redistribute it and/or modify it
! under the terms of the GNU General Public License as published by 
! the Free Software Foundation; either version 2, or (at your option)
! any later version.
! 
! VAMP is distributed in the hope that it will be useful, but
! WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
! GNU General Public License for more details.
! 
! You should have received a copy of the GNU General Public License
! along with this program; if not, write to the Free Software
! Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! This version of the source code of vamp has no comments and
! can be hard to understand, modify, and improve.  You should have
! received a copy of the literate `noweb' sources of vamp that
! contain the documentation in full detail.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
module linalg
  use kinds
  use utils
  implicit none
  private
  public :: lu_decompose
  public :: determinant
  public :: diagonalize_real_symmetric
  private :: jacobi_rotation
  public :: unit, diag
contains
  pure subroutine lu_decompose (a, pivots, eps, l, u)
    real(kind=default), dimension(:,:), intent(inout) :: a
    integer, dimension(:), intent(out), optional :: pivots
    real(kind=default), intent(out), optional :: eps
    real(kind=default), dimension(:,:), intent(out), optional :: l, u
    real(kind=default), dimension(size(a,dim=1)) :: vv
    integer, dimension(size(a,dim=1)) :: p
    integer :: j, pivot
    if (present (eps)) then
       eps = 1.0
    end if
    vv = maxval (abs (a), dim=2)
    if (any (vv == 0.0)) then
       a = 0.0
       if (present (pivots)) then
          pivots = 0
       end if
       if (present (eps)) then
          eps = 0
       end if
       return
    end if
    vv = 1.0 / vv
    do j = 1, size (a, dim=1)
       pivot = j - 1 + sum (maxloc (vv(j:) * abs (a(j:,j))))
       if (j /= pivot) then
          call swap (a(pivot,:), a(j,:))
          if (present (eps)) then
             eps = - eps
          end if
          vv(pivot) = vv(j)
       end if
       p(j) = pivot
       if (a(j,j) == 0.0) then
          a(j,j) = tiny (a(j,j))
       end if
       a(j+1:,j) = a(j+1:,j) / a(j,j)
       a(j+1:,j+1:) &
            = a(j+1:,j+1:) - outer_product (a(j+1:,j), a(j,j+1:))
    end do
    if (present (pivots)) then
       pivots = p
    end if
    if (present (l)) then
       do j = 1, size (a, dim=1)
          l(1:j-1,j) = 0.0
          l(j,j) = 1.0
          l(j+1:,j) = a(j+1:,j)
       end do
       do j = size (a, dim=1), 1, -1
          call swap (l(j,:), l(p(j),:))
       end do
    end if
    if (present (u)) then
       do j = 1, size (a, dim=1)
          u(1:j,j) = a(1:j,j)
          u(j+1:,j) = 0.0
       end do
    end if
  end subroutine lu_decompose
  pure subroutine determinant (a, det)
    real(kind=default), dimension(:,:), intent(in) :: a
    real(kind=default), intent(out) :: det
    real(kind=default), dimension(size(a,dim=1),size(a,dim=2)) :: lu
    integer :: i
    lu = a
    call lu_decompose (lu, eps = det)
    do i = 1, size (a, dim = 1)
       det = det * lu(i,i)
    end do
  end subroutine determinant
  pure subroutine diagonalize_real_symmetric (a, eval, evec, num_rot)
    real(kind=default), dimension(:,:), intent(in) :: a
    real(kind=default), dimension(:), intent(out) :: eval
    real(kind=default), dimension(:,:), intent(out) :: evec
    integer, intent(out), optional :: num_rot
    real(kind=default), dimension(size(a,dim=1),size(a,dim=2)) :: aa
    real(kind=default) :: off_diagonal_norm, threshold, &
         c, g, h, s, t, tau, cot_2phi
    logical, dimension(size(eval),size(eval)) :: upper_triangle
    integer, dimension(size(eval)) :: one_to_ndim
    integer :: p, q, ndim, j, sweep
    integer, parameter :: MAX_SWEEPS = 50
    ndim = size (eval)
    one_to_ndim = (/ (j, j=1,ndim) /)
    upper_triangle = &
         spread (one_to_ndim, dim=1, ncopies=ndim) &
           > spread (one_to_ndim, dim=2, ncopies=ndim)
    aa = a
    call unit (evec)
    if (present (num_rot)) then
       num_rot = 0
    end if
    sweeps: do sweep = 1, MAX_SWEEPS
       off_diagonal_norm = sum (abs (aa), mask=upper_triangle)
       if (off_diagonal_norm == 0.0) then
          eval = diag (aa)
          return
       end if
       if (sweep < 4) then
          threshold = 0.2 * off_diagonal_norm / ndim**2
       else
          threshold = 0.0
       end if
       do p = 1, ndim - 1
          do q = p + 1, ndim
             g = 100 * abs (aa (p,q))
             if ((sweep > 4) &
                  .and. (g <= min (spacing (aa(p,p)), spacing (aa(q,q))))) then
                aa(p,q) = 0.0
             else if (abs (aa(p,q)) > threshold) then
                h = aa(q,q) - aa(p,p)
                if (g <= spacing (h)) then
                   t = aa(p,q) / h
                else
                   cot_2phi = 0.5 * h / aa(p,q)
                   t = sign (1.0_default, cot_2phi) &
                         / (abs (cot_2phi) + sqrt (1.0 + cot_2phi**2))
                end if
                c = 1.0 / sqrt (1.0 + t**2)
                s = t * c
                tau = s / (1.0 + c)
                aa(p,p) = aa(p,p) - t * aa(p,q)
                aa(q,q) = aa(q,q) + t * aa(p,q)
                aa(p,q) = 0.0
                call jacobi_rotation (s, tau, aa(1:p-1,p), aa(1:p-1,q))
                call jacobi_rotation (s, tau, aa(p,p+1:q-1), aa(p+1:q-1,q))
                call jacobi_rotation (s, tau, aa(p,q+1:ndim), aa(q,q+1:ndim))
                call jacobi_rotation (s, tau, evec(:,p), evec(:,q))
                if (present (num_rot)) then
                   num_rot = num_rot + 1
                end if
             end if
          end do
       end do
    end do sweeps
    if (present (num_rot)) then
       num_rot = -1
    end if
  !!! print *, "linalg::diagonalize_real_symmetric: exceeded sweep count"
  end subroutine diagonalize_real_symmetric
  pure subroutine jacobi_rotation (s, tau, vp, vq)
    real(kind=default), intent(in) :: s, tau
    real(kind=default), dimension(:), intent(inout) :: vp, vq
    real(kind=default), dimension(size(vp)) :: vp_tmp
    vp_tmp = vp
    vp = vp - s * (vq     + tau * vp)
    vq = vq + s * (vp_tmp - tau * vq)
  end subroutine jacobi_rotation
  pure subroutine unit (u)
    real(kind=default), dimension(:,:), intent(out) :: u
    integer :: i
    u = 0.0
    do i = 1, min (size (u, dim = 1), size (u, dim = 2))
       u(i,i) = 1.0
    end do
  end subroutine unit
  pure function diag (a) result (d)
    real(kind=default), dimension(:,:), intent(in) :: a
    real(kind=default), dimension(min(size(a,dim=1),size(a,dim=2))) :: d
    integer :: i
    do i = 1, min (size (a, dim = 1), size (a, dim = 2))
       d(i) = a(i,i)
    end do
  end function diag
end module linalg

File: whizard/omega/src/linalg.ml
(* linalg.ml --

   Copyright (C) 1999-2022 by

       Wolfgang Kilian <kilian@physik.uni-siegen.de>
       Thorsten Ohl <ohl@physik.uni-wuerzburg.de>
       Juergen Reuter <juergen.reuter@desy.de>
       with contributions from
       Christian Speckner <cnspeckn@googlemail.com>

   WHIZARD is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   WHIZARD is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  *)

(* This is not a functional implementations, but uses imperative
   array in Fotran style for maximimum speed. *)

exception Singular
exception Not_Square

let copy_matrix a =
  Array.init (Array.length a)
    (fun i -> Array.copy a.(i))

let matmul a b =
  let ni = Array.length a
  and nj = Array.length b.(0)
  and n = Array.length b in
  let ab = Array.make_matrix ni nj 0.0 in
  for i = 0 to pred ni do
    for j = 0 to pred nj do
      for k = 0 to pred n do
	ab.(i).(j) <- ab.(i).(j) +. a.(i).(k) *. b.(k).(j)
      done
    done
  done;
  ab

let matmulv a v =
  let na = Array.length a in
  let nv = Array.length v in
  let v' = Array.make na 0.0 in
  for i = 0 to pred na do
    for j = 0 to pred nv do
      v'.(i) <- v'.(i) +. a.(i).(j) *. v.(j)
    done
  done;
  v'

(*i
let maxval = Array.fold_left max 0.0]

let maxval a : float =
  let x = ref a.(0) in
  for i = 1 to Array.length a - 1 do
    x := max !x a.(i)
  done;
  !x
i*)

let maxabsval a : float =
  let x = ref (abs_float a.(0)) in
  for i = 1 to Array.length a - 1 do
    x := max !x (abs_float a.(i))
  done;
  !x

(*i
let minval = Array.fold_left min 0.0

let minval a : float =
  let x = ref a.(0) in
  for i = 1 to Array.length a - 1 do
    x := min !x a.(i)
  done;
  !x

let maxloc (a : float array) n =
  let n' = ref n
  and max_a : float ref = ref a.(n) in
  for i = succ n to Array.length a - 1 do
    let a_i = a.(i) in
    if a_i > !max_a then begin
      n' := i;
      max_a := a_i
    end
  done;
  !n'

let minloc (a : float array) n =
  let n' = ref n
  and min_a : float ref = ref a.(n) in
  for i = succ n to Array.length a - 1 do
    let a_i = a.(i) in
    if a_i < !min_a then begin
      n' := i;
      min_a := a_i
    end
  done;
  !n'

let rec any' f (a : float array) i =
  if i < 0 then
    false
  else if f a.(i) then
    true
  else
    any' f a (pred i)

let any f a = any' f a (Array.length a - 1)
i*)

(* \thocwmodulesection{$LU$ Decomposition}
   \begin{subequations}
   \label{eq:LU}
   \begin{equation}
     A = LU
   \end{equation}
   In more detail
   \begin{multline}
     \begin{pmatrix}
       a_{00} & a_{01} & \ldots & a_{0(n-1)} \\
       a_{10} & a_{11} & \ldots & a_{1(n-1)} \\
       \vdots & \vdots & \vdots & \vdots \\
       a_{(n-1)0} & a_{(n-1)1} & \ldots & a_{(n-1)(n-1)}
     \end{pmatrix}
     = \\
     \begin{pmatrix}
       1      & 0      & \ldots & 0      \\
       l_{10} & 1      & \ldots & 0      \\
       \vdots & \vdots & \vdots & \vdots \\
       l_{(n-1)0} & l_{(n-1)1} & \ldots & 1
     \end{pmatrix}
     \begin{pmatrix}
       u_{00} & u_{01} & \ldots & u_{0(n-1)} \\
       0      & u_{11} & \ldots & u_{1(n-1)} \\
       \vdots & \vdots & \vdots & \vdots \\
       0      & 0      & \ldots & u_{(n-1)(n-1)}
     \end{pmatrix}
   \end{multline}
   \end{subequations}
   Rewriting~(\ref{eq:LU}) in block matrix notation
   \begin{equation}
     \begin{pmatrix}
       a_{00}     & a_{0\cdot} \\
       a_{\cdot0} & A
     \end{pmatrix}
     =
     \begin{pmatrix}
       1          & 0 \\
       l_{\cdot0} & L
     \end{pmatrix}
     \begin{pmatrix}
       u_{00} & u_{0\cdot} \\
       0      & U
     \end{pmatrix}
      =
     \begin{pmatrix}
       u_{00}            & u_{0\cdot} \\
       l_{\cdot0} u_{00} & l_{\cdot0} \otimes u_{0\cdot} + LU
     \end{pmatrix}
   \end{equation}
   we can solve it easily
   \begin{subequations}
   \begin{align}
     u_{00}     &= a_{00} \\
     u_{0\cdot} &= a_{0\cdot} \\
   \label{eq:LU1}
     l_{\cdot0} &= \frac{a_{\cdot0}}{a_{00}} \\
   \label{eq:LU2}
     LU         &= A - \frac{a_{\cdot0} \otimes a_{0\cdot}}{a_{00}}
   \end{align}
   \end{subequations}
   and~(\ref{eq:LU1}) and~(\ref{eq:LU2}) define a simple iterative
   algorithm if we work from the outside in.  It just remains to add
   pivoting. *)

let swap a i j =
  let a_i = a.(i) in
  a.(i) <- a.(j);
  a.(j) <- a_i

let pivot_column v a n =
  let n' = ref n
  and max_va = ref (v.(n) *. (abs_float a.(n).(n))) in
  for i = succ n to Array.length v - 1 do
    let va_i = v.(i) *. (abs_float a.(i).(n)) in
    if va_i > !max_va then begin
      n' := i;
      max_va := va_i
    end
  done;
  !n'

let lu_decompose_in_place a =
  let n = Array.length a in
  let eps = ref 1
  and pivots = Array.make n 0
  and v =
    try
      Array.init n (fun i ->
	let a_i = a.(i) in
	if Array.length a_i <> n then
	  raise Not_Square;
	1.0 /. (maxabsval a_i))
    with
    | Division_by_zero -> raise Singular in
  for i = 0 to pred n do
    let pivot = pivot_column v a i in
    if pivot <> i then begin
      swap a pivot i;
      eps := - !eps;
      v.(pivot) <- v.(i)
    end;
    pivots.(i) <- pivot;
    let inv_a_ii =
      try 1.0 /. a.(i).(i) with Division_by_zero -> raise Singular in
    for j = succ i to pred n do
      a.(j).(i) <- inv_a_ii *. a.(j).(i)
    done;
    for j = succ i to pred n do
      for k = succ i to pred n do
	a.(j).(k) <- a.(j).(k) -. a.(j).(i) *. a.(i).(k)
      done
    done
  done;
  (pivots, !eps)

let lu_decompose_split a pivots =
  let n = Array.length pivots in
  let l = Array.make_matrix n n 0.0 in
  let u = Array.make_matrix n n 0.0 in
  for i = 0 to pred n do
    l.(i).(i) <- 1.0;
    for j = succ i to pred n do
      l.(j).(i) <- a.(j).(i)
    done
  done;
  for i = pred n downto 0 do
    swap l i pivots.(i)
  done;
  for i = 0 to pred n do
    for j = 0 to i do
      u.(j).(i) <- a.(j).(i)
    done
  done;
  (l, u)

let lu_decompose a =
  let a = copy_matrix a in
  let pivots, _ = lu_decompose_in_place a in
  lu_decompose_split a pivots

let lu_backsubstitute a pivots b =
  let n = Array.length a in
  let nonzero = ref (-1) in
  let b = Array.copy b in
  for i = 0 to pred n do
    let ll = pivots.(i) in
    let b_i = ref (b.(ll)) in
    b.(ll) <- b.(i);
    if !nonzero >= 0 then
      for j = !nonzero to pred i do
	b_i := !b_i -. a.(i).(j) *. b.(j)
      done
    else if !b_i <> 0.0 then
      nonzero := i;
    b.(i) <- !b_i
  done;
  for i = pred n downto 0 do
    let b_i = ref (b.(i)) in
    for j = succ i to pred n do
      b_i := !b_i -. a.(i).(j) *. b.(j)
    done;
    b.(i) <- !b_i /. a.(i).(i)
  done;
  b

let solve_destructive a b =
  let pivot, _ = lu_decompose_in_place a in
  lu_backsubstitute a pivot b

let solve_many_destructive a bs =
  let pivot, _ = lu_decompose_in_place a in
  List.map (lu_backsubstitute a pivot) bs

let solve a b =
  solve_destructive (copy_matrix a) b

let solve_many a bs =
  solve_many_destructive (copy_matrix a) bs

(*i
 *  Local Variables:
 *  mode:caml
 *  indent-tabs-mode:nil
 *  page-delimiter:"^(\\* .*\n"
 *  End:
i*)
